engine/src/test/java/org/camunda/bpm/engine/test/persistence/ConnectionPersistenceExceptionTest.java

```diff
@@ -0,0 +1,119 @@
+import org.assertj.core.api.ThrowableAssert;
+import org.camunda.bpm.engine.IdentityService;
+import org.camunda.bpm.engine.identity.User;
+import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl;
+import org.camunda.bpm.engine.impl.db.sql.DbSqlSessionFactory;
+import org.camunda.bpm.engine.impl.test.RequiredDatabase;
+import org.camunda.bpm.engine.impl.util.PropertiesUtil;
+import org.camunda.bpm.engine.test.ProcessEngineRule;
+import org.camunda.bpm.engine.test.util.ProvidedProcessEngineRule;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Properties;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.camunda.bpm.engine.impl.util.ExceptionUtil.PERSISTENCE_CONNECTION_ERROR_CLASS;
+
+@RequiredDatabase(excludes = { DbSqlSessionFactory.H2 })
+public class ConnectionPersistenceExceptionTest {
+
+  @Rule
+  public ProcessEngineRule engineRule = new ProvidedProcessEngineRule();
+
+  protected IdentityService identityService;
+
+  protected ProcessEngineConfigurationImpl engineConfig;
+
+  protected String resetUrl;
+
+  @Before
+  public void assignServices() {
+    identityService = engineRule.getIdentityService();
+
+    engineConfig = engineRule.getProcessEngineConfiguration();
+
+    resetUrl = ((PooledDataSource) engineConfig.getDataSource()).getUrl();
+  }
+
+  @After
+  public void resetEngine() {
+    ((PooledDataSource) engineConfig.getDataSource()).setUrl(resetUrl);
+    engineRule.getIdentityService().deleteUser("foo");
+  }
+
+  @Test
+  public void shouldFailWithConnectionError() {
+    // given
+    User user = identityService.newUser("foo");
+    identityService.saveUser(user);
+
+    // when
+    SQLException sqlException = provokePersistenceConnectionError();
+
+    // then
+    assertThat(sqlException.getSQLState()).startsWith(PERSISTENCE_CONNECTION_ERROR_CLASS);
+  }
+
+  // helper ////////////////////////////////////////////////////////////////////////////////////////
+
+  protected SQLException provokePersistenceConnectionError() {
+    Properties properties = PropertiesUtil.getProperties("/database.properties");
+    String host = (String) properties.get("database.host");
+    String port = (String) properties.get("database.port");
+
+    String jdbcUrl = resetUrl.replace(host + ":" + port, "not-existing-server:123");
+    ((PooledDataSource) engineConfig.getDataSource()).setUrl(jdbcUrl);
+
+    Iterator<Throwable> exceptionHierarchy = catchExceptionHierarchy(() -> {
+      // when
+      identityService.deleteUser("foo");
+    });
+    exceptionHierarchy.next(); // Top-level engine exception
+    exceptionHierarchy.next(); // MyBatis exception
+    return (SQLException) exceptionHierarchy.next(); // SQLException
```

Comment:
🔧 Alternative (just a suggestion): Catch the throwable, use AssertJ to assert it `hasCause` as expected. This would avoid rebuilding a structure that is already there (the throwable and the recursive causes). Plus, you can assert that we wrap this into the new persistence exception type.

Reply:
Done.

Reply:
👍 This reads nicely after the changes and makes it a lot shorter.